develop/items/custom-data-components.md

```diff
@@ -0,0 +1,267 @@
+---
+title: Custom Data Components
+description: Learn how to add custom data to your items using the new 1.21 component system.
+authors:
+  - Romejanic
+---
+
+# Custom Data Components {#custom-data-components}
+
+As your items grow more complex, you may find yourself needing to store custom data associated with your item. The game allows you to store persistent data within an `ItemStack`, and as of 1.21 the way we do that is by using **Data Components**.
+
+Data Components replace NBT data from previous versions with structured data types (i.e. components) which can be applied to an `ItemStack` to store persistent data about that stack. Data components are namespaced, meaning we can implement our own data components to store custom data about an `ItemStack` and access it later. A full list of the vanilla data components can be found on this [Minecraft wiki page](https://minecraft.wiki/w/Data_component_format#List_of_components).
+
+Along with registering custom components, this page covers the general usage of the components API which also applies to vanilla components. You can see and access the definitions of all vanilla components in the `DataComponentTypes` class.
+
+## Registering a Component {#registering-a-component}
+
+As with anything else in your mod you will need to register your custom component using a `ComponentType`. This component type takes a generic argument containing the type of your component's value. We will be focusing on this in more detail further down.
+
+Choose a sensible class to place this in (e.g. your `ModItems` class from [Creating Your First Item](./first-item.md)), and add the following lines.
```

Comment:
I usually put components in their own class, and I think that's resonable to suggest here too.